pkg/adaptor/cloud/azure/provider.go

```diff
@@ -358,3 +367,52 @@ func (p *azureProvider) deleteNetworkInterface(ctx context.Context, nicName stri
 func (p *azureProvider) Teardown() error {
 	return nil
 }
+
+// Add SelectInstanceType method to select an instance type based on the memory and vcpu requirements
+func (p *azureProvider) selectInstanceType(ctx context.Context, spec cloud.InstanceTypeSpec) (string, error) {
+
+	return cloud.SelectInstanceTypeToUse(spec, p.serviceConfig.InstanceSizeSpecList, p.serviceConfig.InstanceSizes, p.serviceConfig.Size)
+}
+
+// Add a method to populate InstanceSizeSpecList for all the instanceSizes
+// available in Azure
+func (p *azureProvider) updateInstanceSizeSpecList() error {
+
+	// Create a new instance of the Virtual Machine Sizes client
+	vmSizesClient, err := armcompute.NewVirtualMachineSizesClient(p.serviceConfig.SubscriptionId, p.azureClient, nil)
+	if err != nil {
+		return fmt.Errorf("creating VM sizes client: %w", err)
+	}
+	// Get the instance sizes from the service config
+	instanceSizes := p.serviceConfig.InstanceSizes
+
+	// If instanceTypes is empty then populate it with the default instance type
+	if len(instanceSizes) == 0 {
+		instanceSizes = append(instanceSizes, p.serviceConfig.Size)
+	}
+
+	// Create a list of instancesizespec
+	var instanceSizeSpecList []cloud.InstanceTypeSpec
+
+	// TODO: Is there an optimal method for this?
+	// Create NewListPager to iterate over the instance types
+	pager := vmSizesClient.NewListPager(p.serviceConfig.Region, &armcompute.VirtualMachineSizesClientListOptions{})
+
+	// Iterate over the page and populate the instanceSizeSpecList for all the instanceSizes
+	for pager.More() {
+		nextResult, err := pager.NextPage(context.Background())
+		if err != nil {
+			return fmt.Errorf("getting next page of VM sizes: %w", err)
+		}
+		for _, vmSize := range nextResult.VirtualMachineSizeListResult.Value {
+			if util.Contains(instanceSizes, *vmSize.Name) {
+				instanceSizeSpecList = append(instanceSizeSpecList, cloud.InstanceTypeSpec{InstanceType: *vmSize.Name, VCPUs: int64(*vmSize.NumberOfCores), Memory: int64(*vmSize.MemoryInMB)})
```

Comment:
Sorry if it has been discussed before, what if multiple instance types matches same {vCPU, Mem} set?

Reply:
The first one in the list should get selected. We can try this scenario by updating the unit tests of the common cloud function to select bestfit instance type.